File: app/src/main/java/com/example/data/service/LinkIngestionService.kt
package com.example.data.service

import com.example.data.models.SourceType
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request
import java.net.URI
import java.util.Locale
import java.util.concurrent.TimeUnit
import java.util.regex.Pattern

data class LinkValidationResult(val isValid: Boolean, val error: String? = null)

data class IngestedContent(
    val transcriptText: String,
    val detectedTitle: String? = null,
    val sourceType: SourceType,
    val sourceUrl: String,
    val durationSeconds: Long = 0L
)

class LinkIngestionService(
    private val client: OkHttpClient = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .build()
) {
    companion object {
        private val PRIVATE_IP_PATTERNS = listOf(
            Pattern.compile("^localhost$", Pattern.CASE_INSENSITIVE),
            Pattern.compile("^127\\.\\d+\\.\\d+\\.\\d+$"),
            Pattern.compile("^0\\.0\\.0\\.0$"),
            Pattern.compile("^::1$"),
            Pattern.compile("^\\[::1\\]$"),
            Pattern.compile("^10\\.\\d+\\.\\d+\\.\\d+$"),
            Pattern.compile("^192\\.168\\.\\d+\\.\\d+$"),
            Pattern.compile("^172\\.(1[6-9]|2\\d|3[0-1])\\.\\d+\\.\\d+$"),
            Pattern.compile("^169\\.254\\.\\d+\\.\\d+$"), // Link-local / metadata services
            Pattern.compile(".*\\.local$", Pattern.CASE_INSENSITIVE),
            Pattern.compile(".*\\.internal$", Pattern.CASE_INSENSITIVE),
            Pattern.compile(".*\\.lan$", Pattern.CASE_INSENSITIVE)
        )
    }

    /**
     * Validates URL syntax and enforces SSRF protection against internal/private endpoints.
     */
    fun validateUrl(urlString: String?): LinkValidationResult {
        if (urlString.isNullOrBlank()) {
            return LinkValidationResult(false, "URL is required and cannot be empty.")
        }

        val uri = try {
            val trimmed = urlString.trim()
            URI.create(trimmed)
        } catch (e: Exception) {
            return LinkValidationResult(false, "Invalid URL format. Please provide a full URL starting with http:// or https://.")
        }

        val scheme = uri.scheme?.lowercase(Locale.ROOT)
        if (scheme != "http" && scheme != "https") {
            return LinkValidationResult(false, "Unsupported protocol: '$scheme'. Only http:// and https:// links are supported.")
        }

        val host = uri.host?.lowercase(Locale.ROOT) ?: ""
        if (host.isBlank()) {
            return LinkValidationResult(false, "URL is missing a valid host.")
        }

        val isPrivate = host == "localhost" ||
                host == "[::1]" ||
                host == "::1" ||
                host.startsWith("127.") ||
                host.contains("169.254") ||
                host == "metadata.google.internal" ||
                PRIVATE_IP_PATTERNS.any { it.matcher(host).matches() }

        if (isPrivate) {
            return LinkValidationResult(
                false,
                "SSRF Security Guard: Requests to local, internal, or private network addresses are strictly prohibited."
            )
        }

        return LinkValidationResult(true)
    }

    /**
     * Classifies URL into SourceType
     */
    fun detectSourceType(urlString: String): SourceType {
        val validation = validateUrl(urlString)
        if (!validation.isValid) {
            throw IllegalArgumentException(validation.error ?: "Invalid URL")
        }

        val uri = URI.create(urlString.trim())
        val host = uri.host?.lowercase(Locale.ROOT) ?: ""
        val path = uri.path?.lowercase(Locale.ROOT) ?: ""

        if (host.contains("youtube.com") || host.contains("youtu.be") ||
            host.contains("vimeo.com") || host.contains("dailymotion.com")) {
            return SourceType.URL_VIDEO
        }

        if (path.endsWith(".mp3") || path.endsWith(".m4a") ||
            path.endsWith(".wav") || path.endsWith(".aac") ||
            path.endsWith(".ogg") || host.contains("podcasts.google.com") ||
            host.contains("podcasts.apple.com") || host.contains("spotify.com") ||
            host.contains("soundcloud.com") || host.contains("podcast") ||
            path.contains("podcast")) {
            return SourceType.URL_AUDIO
        }

        return SourceType.URL_ARTICLE
    }

    /**
     * Ingests content from a URL.
     * Path A (Video/Audio) or Path B (Web Article)
     * Strictly fails with typed exceptions on unreachable/paywalled/unsupported sources.
     */
    suspend fun ingestUrl(
        url: String,
        onProgress: (String) -> Unit = {}
    ): IngestedContent = withContext(Dispatchers.IO) {
        val validation = validateUrl(url)
        if (!validation.isValid) {
            throw IllegalArgumentException(validation.error ?: "Invalid URL for ingestion.")
        }

        val sourceType = detectSourceType(url)
        onProgress("Connecting to $url...")

        val request = Request.Builder()
            .url(url)
            .header("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) LectureNotesAI/1.0")
            .build()

        val response = try {
            client.newCall(request).execute()
        } catch (e: Exception) {
            throw IllegalStateException("UNREACHABLE: Network request failed: ${e.localizedMessage}")
        }

        response.use { resp ->
            if (!resp.isSuccessful) {
                when (resp.code) {
                    401, 403 -> throw IllegalStateException("PAYWALLED: This resource requires login or subscription (HTTP ${resp.code}).")
                    404 -> throw IllegalStateException("UNREACHABLE: Resource was not found (HTTP 404).")
                    else -> throw IllegalStateException("UNREACHABLE: Failed to fetch link content (HTTP ${resp.code}).")
                }
            }

            val contentType = resp.header("Content-Type", "")?.lowercase(Locale.ROOT) ?: ""
            if (!contentType.contains("text/html") && !contentType.contains("text/plain") && !contentType.contains("audio/")) {
                throw IllegalStateException("UNSUPPORTED_TYPE: URL did not return HTML, text, or audio content (Content-Type: $contentType).")
            }

            val bodyString = resp.body?.string()
                ?: throw IllegalStateException("Response body was empty.")

            onProgress("Extracting lecture text and stripping boilerplate...")
            val (title, text) = extractArticleText(bodyString)

            if (text.isBlank()) {
                throw IllegalStateException("Article extraction yielded no readable text (empty page content). The page may require JavaScript or authentication.")
            }

            val estimatedDuration = (text.split(Regex("\\s+")).filter { it.isNotBlank() }.size * 0.3).toLong().coerceAtLeast(60L)

            return@withContext IngestedContent(
                transcriptText = text,
                detectedTitle = title,
                sourceType = sourceType,
                sourceUrl = url,
                durationSeconds = estimatedDuration
            )
        }
    }

    /**
     * Pure HTML text extractor: Strips boilerplate, navigation, scripts, ads, and footers.
     */
    fun extractArticleText(html: String): Pair<String?, String> {
        if (html.isBlank()) return Pair(null, "")

        // Extract title
        var title: String? = null
        val ogTitleMatcher = Pattern.compile("<meta[^>]+property=[\"']og:title[\"'][^>]+content=[\"']([^\"']+)[\"']", Pattern.CASE_INSENSITIVE).matcher(html)
        if (ogTitleMatcher.find()) {
            title = ogTitleMatcher.group(1)?.trim()
        } else {
            val titleMatcher = Pattern.compile("<title[^>]*>([^<]+)</title>", Pattern.CASE_INSENSITIVE).matcher(html)
            if (titleMatcher.find()) {
                title = titleMatcher.group(1)?.trim()
            }
        }

        val cleaned = html
            .replace(Regex("(?i)<script\\b[^<]*(?:(?!</script>)<[^<]*)*</script>"), " ")
            .replace(Regex("(?i)<style\\b[^<]*(?:(?!</style>)<[^<]*)*</style>"), " ")
            .replace(Regex("(?i)<noscript\\b[^<]*(?:(?!</noscript>)<[^<]*)*</noscript>"), " ")
            .replace(Regex("(?i)<header\\b[^<]*(?:(?!</header>)<[^<]*)*</header>"), " ")
            .replace(Regex("(?i)<footer\\b[^<]*(?:(?!</footer>)<[^<]*)*</footer>"), " ")
            .replace(Regex("(?i)<nav\\b[^<]*(?:(?!</nav>)<[^<]*)*</nav>"), " ")
            .replace(Regex("(?i)<aside\\b[^<]*(?:(?!</aside>)<[^<]*)*</aside>"), " ")
            .replace(Regex("<[^>]+>"), " ")
            .replace("&nbsp;", " ")
            .replace("&amp;", "&")
            .replace("&lt;", "<")
            .replace("&gt;", ">")
            .replace("&quot;", "\"")

        val normalized = cleaned
            .replace("\r\n", "\n")
            .replace("\t", " ")
            .replace(Regex("[ \t]{2,}"), " ")
            .replace(Regex("\n{3,}"), "\n\n")
            .trim()

        return Pair(title, normalized)
    }
}

File: src/services/linkIngestionService.ts
import { Source, SourceType, ValidationCheckResult } from '../types';

export interface LinkDetectionResult {
  sourceType: SourceType;
  platform: string;
  normalizedUrl: string;
  hostname: string;
}

export interface IngestedContentResult {
  success: boolean;
  transcriptText?: string;
  textCandidate?: string;
  audioBlob?: Blob | null;
  detectedTitle?: string;
  title?: string;
  sourceType: SourceType;
  sourceUrl: string;
  durationSeconds?: number;
  error?: string;
}

const PRIVATE_IP_PATTERNS = [
  /^localhost$/i,
  /^127\.\d+\.\d+\.\d+$/,
  /^0\.0\.0\.0$/,
  /^::1$/,
  /^\[::1\]$/,
  /^10\.\d+\.\d+\.\d+$/,
  /^192\.168\.\d+\.\d+$/,
  /^172\.(1[6-9]|2\d|3[0-1])\.\d+\.\d+$/,
  /^169\.254\.\d+\.\d+$/, // Link-local / metadata services
  /\.local$/i,
  /\.internal$/i,
  /\.lan$/i
];

export class LinkIngestionService {
  /**
   * Validates URL syntax and enforces SSRF protection against internal/private endpoints.
   */
  validateUrl(urlString: string): ValidationCheckResult {
    if (!urlString || urlString.trim().length === 0) {
      return { valid: false, error: 'URL is required and cannot be empty.' };
    }

    let parsed: URL;
    try {
      parsed = new URL(urlString.trim());
    } catch {
      return { valid: false, error: 'Invalid URL format. Please provide a full URL starting with http:// or https://.' };
    }

    if (parsed.protocol !== 'http:' && parsed.protocol !== 'https:') {
      return { valid: false, error: `Unsupported protocol: '${parsed.protocol}'. Only http:// and https:// links are supported.` };
    }

    const hostname = parsed.hostname.toLowerCase();

    // SSRF Guard (§12)
    const isPrivate = PRIVATE_IP_PATTERNS.some(pat => pat.test(hostname));
    if (isPrivate) {
      return {
        valid: false,
        error: 'SSRF Security Guard: Requests to local, internal, or private network addresses are strictly prohibited.'
      };
    }

    // Check for suspicious internal metadata services (e.g. AWS 169.254.169.254)
    if (hostname.includes('169.254') || hostname === 'metadata.google.internal') {
      return {
        valid: false,
        error: 'SSRF Security Guard: Access to cloud metadata services is blocked.'
      };
    }

    return { valid: true };
  }

  /**
   * Classifies the target URL into one of the supported source types.
   */
  detectSourceType(urlString: string): LinkDetectionResult {
    const valid = this.validateUrl(urlString);
    if (!valid.valid) {
      throw new Error(valid.error || 'Invalid URL');
    }

    const parsed = new URL(urlString.trim());
    const host = parsed.hostname.toLowerCase();
    const pathname = parsed.pathname.toLowerCase();

    // Video platforms
    if (
      host.includes('youtube.com') ||
      host.includes('youtu.be') ||
      host.includes('vimeo.com') ||
      host.includes('dailymotion.com')
    ) {
      return {
        sourceType: 'URL_VIDEO',
        platform: host.includes('youtu') ? 'YouTube' : 'Video Platform',
        normalizedUrl: parsed.href,
        hostname: host
      };
    }

    // Direct audio files or podcast audio
    if (
      pathname.endsWith('.mp3') ||
      pathname.endsWith('.m4a') ||
      pathname.endsWith('.wav') ||
      pathname.endsWith('.aac') ||
      pathname.endsWith('.ogg') ||
      host.includes('podcasts.google.com') ||
      host.includes('podcasts.apple.com') ||
      host.includes('spotify.com') ||
      host.includes('soundcloud.com') ||
      host.includes('podcast') ||
      pathname.includes('podcast')
    ) {
      return {
        sourceType: 'URL_AUDIO',
        platform: 'Audio / Podcast',
        normalizedUrl: parsed.href,
        hostname: host
      };
    }

    // Standard web article / lecture notes page
    return {
      sourceType: 'URL_ARTICLE',
      platform: 'Web Article',
      normalizedUrl: parsed.href,
      hostname: host
    };
  }

  /**
   * Ingests a URL according to Path A (Audio/Video) or Path B (Web Article).
   * Strictly outputs verbatim text candidate or raw audio; NEVER creates notes or summaries.
   */
  async ingestUrl(
    sourceOrUrl: Source | string,
    onProgress?: (msg: string) => void
  ): Promise<IngestedContentResult> {
    const url = typeof sourceOrUrl === 'string' ? sourceOrUrl : (sourceOrUrl.sourceUrl || '');
    const validation = this.validateUrl(url);
    if (!validation.valid) {
      return {
        success: false,
        error: validation.error || 'Invalid URL for ingestion.',
        sourceType: 'URL_ARTICLE',
        sourceUrl: url
      };
    }

    const detection = this.detectSourceType(url);
    const source: Source = typeof sourceOrUrl === 'string'
      ? {
          id: `src_temp_${Date.now()}`,
          sourceType: detection.sourceType,
          sourceUrl: url,
          userId: 'usr_101',
          title: detection.platform,
          subject: 'General',
          durationSeconds: 0,
          createdAt: Date.now(),
          status: 'FETCHING'
        }
      : sourceOrUrl;

    if (onProgress) onProgress(`Connecting to ${detection.platform} (${detection.hostname})...`);

    try {
      let res: IngestedContentResult;
      if (detection.sourceType === 'URL_VIDEO') {
        res = await this.ingestVideo(source, detection, onProgress);
      } else if (detection.sourceType === 'URL_AUDIO') {
        res = await this.ingestAudio(source, detection, onProgress);
      } else {
        res = await this.ingestArticle(source, detection, onProgress);
      }
      return {
        ...res,
        success: true,
        title: res.title || res.detectedTitle,
        transcriptText: res.transcriptText || res.textCandidate
      };
    } catch (err: unknown) {
      const errorMsg = err instanceof Error ? err.message : 'Failed to ingest URL.';
      return {
        success: false,
        error: errorMsg,
        sourceType: detection.sourceType,
        sourceUrl: url
      };
    }
  }

  /**
   * Path A (Video): Fetches video metadata or transcript tracks.
   */
  private async ingestVideo(
    source: Source,
    detection: LinkDetectionResult,
    onProgress?: (msg: string) => void
  ): Promise<IngestedContentResult> {
    const url = source.sourceUrl || '';
    const parsed = new URL(url);

    if (onProgress) onProgress('Resolving video lecture captions and speech metadata...');

    // Extract YouTube Video ID
    let videoId: string | null = null;
    if (parsed.hostname.includes('youtu.be')) {
      videoId = parsed.pathname.replace(/^\//, '').split('/')[0];
    } else if (parsed.hostname.includes('youtube.com')) {
      videoId = parsed.searchParams.get('v');
    }

    if (!videoId && detection.platform === 'YouTube') {
      throw new Error('Could not identify a valid YouTube video ID from the provided link.');
    }

    // Attempt to fetch public caption or audio metadata
    try {
      if (onProgress) onProgress('Fetching lecture transcript from video source...');

      // In client web apps, direct scraping of YouTube watch page or caption tracks can hit CORS if not proxied.
      // We perform a safe fetch attempt, and if CORS blocked or no captions, provide clear guidance without fabrication.
      const response = await fetch(url, { method: 'GET', mode: 'cors' }).catch(() => null);

      if (response && response.ok) {
        const html = await response.text();
        const extracted = this.extractArticleText(html);
        if (extracted.text.length > 50) {
          return {
            success: true,
            textCandidate: extracted.text,
            transcriptText: extracted.text,
            detectedTitle: extracted.title || source.title,
            title: extracted.title || source.title,
            sourceType: 'URL_VIDEO',
            sourceUrl: url,
            durationSeconds: source.durationSeconds || 1800
          };
        }
      }

      // If YouTube or video platform restricts direct browser-side HTML extraction,
      // provide a clean typed failure or instructions for video transcripts
      throw new Error(
        `Direct client-side extraction for ${detection.platform} video is restricted by CORS/platform policies. ` +
        `Please provide a direct audio/podcast URL, an article link, or transcribe using the microphone.`
      );
    } catch (err: unknown) {
      const msg = err instanceof Error ? err.message : 'Failed to retrieve video lecture transcript.';
      throw new Error(msg);
    }
  }

  /**
   * Path A (Audio): Downloads direct audio stream (mp3/m4a/wav) for the TranscriptionService.
   */
  private async ingestAudio(
    source: Source,
    detection: LinkDetectionResult,
    onProgress?: (msg: string) => void
  ): Promise<IngestedContentResult> {
    const url = source.sourceUrl || '';
    if (onProgress) onProgress(`Streaming audio track from ${detection.hostname}...`);

    try {
      const response = await fetch(url);
      if (!response.ok) {
        if (response.status === 401 || response.status === 403) {
          throw new Error(`PAYWALLED: The audio source requires authentication or payment (HTTP ${response.status}).`);
        }
        throw new Error(`UNREACHABLE: Audio stream could not be reached (HTTP ${response.status}).`);
      }

      const audioBlob = await response.blob();

      if (!audioBlob || audioBlob.size === 0) {
        throw new Error('Audio file at URL returned 0 bytes.');
      }

      return {
        success: true,
        audioBlob,
        detectedTitle: source.title,
        title: source.title,
        sourceType: 'URL_AUDIO',
        sourceUrl: url,
        durationSeconds: source.durationSeconds || 1800
      };
    } catch (err: unknown) {
      const msg = err instanceof Error ? err.message : 'Failed to download audio track from URL.';
      throw new Error(msg);
    }
  }

  /**
   * Path B (Article): Fetches webpage HTML and extracts substantive lecture text.
   */
  private async ingestArticle(
    source: Source,
    detection: LinkDetectionResult,
    onProgress?: (msg: string) => void
  ): Promise<IngestedContentResult> {
    const url = source.sourceUrl || '';
    if (onProgress) onProgress(`Fetching article page from ${detection.hostname}...`);

    let html = '';
    try {
      const response = await fetch(url);
      if (!response.ok) {
        if (response.status === 401 || response.status === 403) {
          throw new Error(`PAYWALLED: This article requires login or subscription (HTTP ${response.status}).`);
        }
        if (response.status === 404) {
          throw new Error('UNREACHABLE: Article page was not found (HTTP 404).');
        }
        throw new Error(`UNREACHABLE: Could not fetch article from link (HTTP ${response.status}).`);
      }

      const contentType = response.headers.get('content-type') || '';
      if (!contentType.includes('text/html') && !contentType.includes('text/plain')) {
        throw new Error(`UNSUPPORTED_TYPE: URL did not return HTML or text content (Content-Type: ${contentType}).`);
      }

      html = await response.text();
    } catch (fetchErr: unknown) {
      const msg = fetchErr instanceof Error ? fetchErr.message : 'Network error fetching article link.';
      throw new Error(msg);
    }

    if (onProgress) onProgress('Extracting readable lecture content and stripping boilerplate...');
    const extracted = this.extractArticleText(html);

    if (!extracted.text || extracted.text.trim().length === 0) {
      throw new Error('Article extraction yielded no readable text (empty page content). The page may require JavaScript or authentication.');
    }

    return {
      success: true,
      textCandidate: extracted.text,
      transcriptText: extracted.text,
      detectedTitle: extracted.title || source.title,
      title: extracted.title || source.title,
      sourceType: 'URL_ARTICLE',
      sourceUrl: url,
      durationSeconds: Math.max(60, Math.round(extracted.text.split(/\s+/).length / 2.5)) // rough reading time equivalent
    };
  }

  /**
   * Pure HTML text extractor: Strips boilerplate, navigation, scripts, ads, and footers.
   */
  extractArticleText(html: string): { title?: string; text: string } {
    if (!html) return { text: '' };

    // Extract Title
    let title: string | undefined;
    const ogTitleMatch = html.match(/<meta[^>]+property=["']og:title["'][^>]+content=["']([^"']+)["']/i);
    if (ogTitleMatch) {
      title = ogTitleMatch[1].trim();
    } else {
      const titleMatch = html.match(/<title[^>]*>([^<]+)<\/title>/i);
      if (titleMatch) {
        title = titleMatch[1].trim();
      }
    }

    // Clean HTML string using regular expressions for environments without DOM,
    // or standard DOMParser if running in browser
    let cleanText = '';

    if (typeof window !== 'undefined' && typeof window.DOMParser !== 'undefined') {
      try {
        const parser = new DOMParser();
        const doc = parser.parseFromString(html, 'text/html');

        // Remove scripts, styles, navigations, footers, and ads
        const removeSelectors = [
          'script',
          'style',
          'noscript',
          'iframe',
          'nav',
          'header',
          'footer',
          'aside',
          'form',
          '.advertisement',
          '.ad',
          '.social-share',
          '.comments'
        ];

        removeSelectors.forEach(sel => {
          doc.querySelectorAll(sel).forEach(el => el.remove());
        });

        // Target main article body if available
        const main = doc.querySelector('main, article, [role="main"], .post-content, .article-content, #content') || doc.body;
        cleanText = (main ? main.textContent : doc.body.textContent) || '';
      } catch {
        cleanText = this.stripTagsRegex(html);
      }
    } else {
      cleanText = this.stripTagsRegex(html);
    }

    // Meaning-preserving whitespace normalization
    const normalized = cleanText
      .replace(/\r\n/g, '\n')
      .replace(/\t/g, ' ')
      .replace(/[ \t]{2,}/g, ' ')
      .replace(/\n{3,}/g, '\n\n')
      .trim();

    return { title, text: normalized };
  }

  private stripTagsRegex(html: string): string {
    return html
      .replace(/<script\b[^<]*(?:(?!<\/script>)<[^<]*)*<\/script>/gi, ' ')
      .replace(/<style\b[^<]*(?:(?!<\/style>)<[^<]*)*<\/style>/gi, ' ')
      .replace(/<noscript\b[^<]*(?:(?!<\/noscript>)<[^<]*)*<\/noscript>/gi, ' ')
      .replace(/<header\b[^<]*(?:(?!<\/header>)<[^<]*)*<\/header>/gi, ' ')
      .replace(/<footer\b[^<]*(?:(?!<\/footer>)<[^<]*)*<\/footer>/gi, ' ')
      .replace(/<nav\b[^<]*(?:(?!<\/nav>)<[^<]*)*<\/nav>/gi, ' ')
      .replace(/<aside\b[^<]*(?:(?!<\/aside>)<[^<]*)*<\/aside>/gi, ' ')
      .replace(/<[^>]+>/g, ' ')
      .replace(/&nbsp;/g, ' ')
      .replace(/&amp;/g, '&')
      .replace(/&lt;/g, '<')
      .replace(/&gt;/g, '>')
      .replace(/&quot;/g, '"');
  }
}

export const linkIngestionService = new LinkIngestionService();
